{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'flair'...\n",
      "remote: Enumerating objects: 11, done.\u001b[K\n",
      "remote: Counting objects: 100% (11/11), done.\u001b[K\n",
      "remote: Compressing objects: 100% (10/10), done.\u001b[K\n",
      "remote: Total 628 (delta 1), reused 5 (delta 1), pack-reused 617\u001b[K\n",
      "Receiving objects: 100% (628/628), 19.27 MiB | 6.89 MiB/s, done.\n",
      "Resolving deltas: 100% (316/316), done.\n",
      "Checking connectivity... done.\n",
      "Checking out files: 100% (44/44), done.\n"
     ]
    }
   ],
   "source": [
    "!git clone https://github.com/BrooksLabUCSC/flair.git\n",
    "!pip install kerneltree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import csv\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Script for Flair. 1) Converts bam mapped reads file to bed12 and corrects misaligned splice sites using a genome annotation. 2) Collapse corrected reads into high-confidence isoforms and quantifies them by counting converted into tpm.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Flair(\n",
    "    main_dir, #full path of working directory\n",
    "    bam_input, #only filename, path same as main_dir\n",
    "    genome,    #full path of reference genome (in ref_fasta folder)\n",
    "    chr_sizes, #full path to tab-delim file of chromosome sizes, names must correspond to both annotation and reference\n",
    "    chr_annot, #full path to gtf annotation file\n",
    "    reads,     #full path to fa file of UN-mapped raw reads\n",
    "    file_org): #full path to tab-delim file listing sample information\n",
    "    \n",
    "    !mkdir {main_dir}/nc.flair\n",
    "#     !samtools faidx {genome}  <- if using a new genome fasta file\n",
    "    print(\"Copy paste into command line window\")\n",
    "    print(\"1 convert, correct, collapse scripts:\") ##**NB: module tqdm only installed on python3, not just python\n",
    "    #both samtools and minimap2 directories are located in software/flair/\n",
    "        #-s parameter is minimum number of supporting reads for an isoform (default: 3) \n",
    "        #--stringent\n",
    "    print('$python3 software/flair/bin/bam2Bed12.py -i '+main_dir+'/'+bam_input+' > '+ main_dir+'/nc.flair/_clean_sorted.bed12;python3 software/flair/flair.py correct -c '+chr_sizes+' -g '+genome+' -q '+main_dir+'/nc.flair/_clean_sorted.bed12 -o '+main_dir+'/nc.flair/flair -f '+chr_annot+'; time python3 software/flair/flair.py collapse -g '+genome+' -r '+reads+' -f '+chr_annot+' -o '+main_dir+\"/nc.flair/flair.collapse -q \"+main_dir+'/nc.flair/flair_all_corrected.psl')\n",
    "    main_dir=main_dir+'/nc.flair'\n",
    "    print \"2 quantify:\"\n",
    "    print 'python3 software/flair/flair.py quantify --tpm -r '+file_org+' -o '+main_dir+'/counts_matrix.tsv -i '+main_dir+'/flair.collapse.isoforms.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘2019-8-6_cdna/a431.b1.merged/nc.flair’: File exists\n",
      "Copy paste into command line window\n",
      "1 convert, correct, collapse scripts:\n",
      "$python3 software/flair/bin/bam2Bed12.py -i 2019-8-6_cdna/a431.b1.merged/_cleansort.bam > 2019-8-6_cdna/a431.b1.merged/nc.flair/_clean_sorted.bed12;python3 software/flair/flair.py correct -c software/bedtools2/genomes/clean.human.hg38.genome.chrom.sizes.tsv -g 2019-6-13_A431/GRCh38_trimmed.fa -q 2019-8-6_cdna/a431.b1.merged/nc.flair/_clean_sorted.bed12 -o 2019-8-6_cdna/a431.b1.merged/nc.flair/flair -f software/flair/new.gencode.v29.annotation.gtf; time python3 software/flair/flair.py collapse -g 2019-6-13_A431/GRCh38_trimmed.fa -r 2019-8-6_cdna/a431.b1/a431.b1.fasta -f software/flair/new.gencode.v29.annotation.gtf -o 2019-8-6_cdna/a431.b1.merged/nc.flair/flair.collapse -q 2019-8-6_cdna/a431.b1.merged/nc.flair/flair_all_corrected.psl\n",
      "2 quantify:\n",
      "python3 software/flair/flair.py quantify --tpm -r 2019-8-6_cdna/a431.b1.merged/nc.flair/ctmouse1_dna.txt -o 2019-8-6_cdna/a431.b1.merged/nc.flair/counts_matrix.tsv -i 2019-8-6_cdna/a431.b1.merged/nc.flair/flair.collapse.isoforms.fa\n"
     ]
    }
   ],
   "source": [
    "Flair(\n",
    "    main_dir='2019-8-6_cdna/a431.b1.merged',\n",
    "    bam_input='_cleansort.bam',\n",
    "    genome='software/ref_fasta/homo_GRCh38_trimmed_ref.fa',\n",
    "    chr_sizes='software/bedtools2/genomes/clean.human.hg38.genome.chrom.sizes.tsv',\n",
    "    chr_annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "    reads='2019-8-6_cdna/a431.b1/a431.b1.fasta',   ##use raw, unaligned reads\n",
    "    file_org='2019-7-26_cdna.txt'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Isoform productivity: append label to end of each line of PSL\n",
    "    ##0: functional stop codon (productive)\n",
    "    ##1: premature stop codon (unproductive)\n",
    "    ##2: lncRNA (no start codon) \n",
    "    \n",
    "    #From script: 'Unproductive proportion estimate: unproductive / (valid_transcripts + unproductive))\n",
    "    ##Single-exon transcripts are counted as lncRNAs.\n",
    "\n",
    "def IsoformProductivity(directory,\n",
    "                       reads,\n",
    "                       annot,\n",
    "                       genome):\n",
    "    reads=directory+reads\n",
    "    outfile=directory+'/isoforms.productivity.psl'\n",
    "#     outfile=directory+'/isosproductive.psl'\n",
    "    !python software/flair/bin/mark_productivity.py {reads} {annot} {genome} > {outfile}\n",
    "\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "\n",
    "    good_stop=0\n",
    "    bad_stop=0\n",
    "    no_start=0\n",
    "    single_exon=0\n",
    "    for row in reader:\n",
    "        n=int(row[-1])\n",
    "        if n==0: good_stop+=1\n",
    "        elif n==1: bad_stop+=1\n",
    "        elif n==2: no_start+=1\n",
    "        if int(row[17])==1: single_exon+=1\n",
    "\n",
    "    print directory        \n",
    "    print 'Productive isoforms: '+str(good_stop)\n",
    "    print 'Unproductive isoforms: '+str(bad_stop)\n",
    "    total=good_stop+bad_stop+no_start\n",
    "    print 'Proportion of productive: ' +str(good_stop*100 / (total))+'%'\n",
    "    print 'lncRNA/other: '+str(no_start - single_exon)\n",
    "    print 'Single exon transcripts: '+str(single_exon)\n",
    "    Productive_isos(directory,outfile)\n",
    "\n",
    "def Productive_isos(directory,\n",
    "                   infile):\n",
    "    out=directory+'/all_productive_isos.psl'\n",
    "    i=open(infile,'r')\n",
    "    o=open(out,'a')\n",
    "    reader=csv.reader(i,delimiter='\\t')\n",
    "    writer = csv.writer(o,delimiter='\\t')\n",
    "    for row in reader:\n",
    "        length=abs(int(row[16])-int(row[15]))\n",
    "        if int(row[-1])==2 and length>=300: writer.writerow(row)\n",
    "        elif int(row[21])==0: writer.writerow(row)   \n",
    "\n",
    "##Marks each isoform by appending new column to each entry of PSL\n",
    "    ##0: spliced\n",
    "    ##1: intron-retaining\n",
    "    \n",
    "def IntronRetention(directory,\n",
    "                   isoforms='exonsgenes.collapse.isoforms.psl',\n",
    "                   outfile='isoforms.ir.psl'):\n",
    "    isoforms=directory+'/'+isoforms \n",
    "    outfile=directory+'/'+outfile\n",
    "    coords=directory+'/coords.txt'\n",
    "    !python software/flair/bin/mark_intron_retention.py {isoforms} {outfile} {coords}\n",
    "\n",
    "    ##Count intron-retaining and spliced isoforms\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    spliced=0\n",
    "    intron_retain=0\n",
    "    splice_pos=0\n",
    "    splice_neg=0\n",
    "    retain_pos=0\n",
    "    retain_neg=0\n",
    "    for row in reader:\n",
    "        if int(row[21])==0: spliced+=1\n",
    "        elif int(row[21])==1: intron_retain+=1\n",
    "\n",
    "    #Count spliced isoforms based on strand\n",
    "        if int(row[21])==0 and row[8] is '+': splice_pos+=1\n",
    "        elif int(row[21])==0 and row[8] is '-': splice_neg+=1\n",
    "        if int(row[21])==1 and row[8] is '+': retain_pos+=1\n",
    "        elif int(row[21])==1 and row[8] is '-': retain_neg+=1    \n",
    "\n",
    "    print directory\n",
    "    print 'Spliced isoforms: '+str(spliced)\n",
    "    print 'Intron-retaining isoforms: '+str(intron_retain)\n",
    "    print '+ strand, spliced: '+str(splice_pos)\n",
    "    print '- strand, spliced: '+str(splice_neg)\n",
    "    print '+ strand, retaining: '+str(retain_pos)\n",
    "    print '+ strand, retaining: '+str(retain_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unproductive proportion estimate 0.154203851574\n",
      "2019-8-6_cdna/a431.b1/trcl_VA/nc.flair\n",
      "Productive isoforms: 14724\n",
      "Unproductive isoforms: 3283\n",
      "Proportion of productive: 33%\n",
      "lncRNA/other: 5316\n",
      "Single exon transcripts: 20093\n"
     ]
    }
   ],
   "source": [
    "IsoformProductivity('2019-8-6_cdna/a431.b1/trcl_VA/nc.flair',\n",
    "                       reads='/flair.collapse.isoforms.psl',\n",
    "                       annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "                       genome='2019-6-13_A431/GRCh38_trimmed.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-7-22_dirrna/trcl_VA/flair\n",
      "Spliced isoforms: 3060\n",
      "Intron-retaining isoforms: 0\n",
      "+ strand, spliced: 1603\n",
      "- strand, spliced: 1457\n",
      "+ strand, retaining: 0\n",
      "+ strand, retaining: 0\n"
     ]
    }
   ],
   "source": [
    "IntronRetention('2019-7-22_dirrna/trcl_VA/flair',\n",
    "                   isoforms='isosproductive.psl',\n",
    "                   outfile='isosproductive.ir.psl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8588\n",
      "longest single exon: 6696\n",
      "8801\n",
      "longest single exon: 8062\n"
     ]
    }
   ],
   "source": [
    "#debug code to determine longest single exon transcript\n",
    "\n",
    "f=open('2019-7-26_cdna/a431utc/trcl_VA/flair/isoforms.productivity.psl')\n",
    "reader=csv.reader(f,delimiter='\\t')\n",
    "count=0\n",
    "longest=0\n",
    "printrow=''\n",
    "for row in reader:\n",
    "    if int(row[-1])==2 and int(row[0])>=300: \n",
    "        count+=1\n",
    "        if int(row[0])>longest: longest=int(row[0])\n",
    "\n",
    "print count\n",
    "print 'longest single exon: '+str(longest)\n",
    "f.close()\n",
    "        \n",
    "outfile='2019-7-26_cdna/a431.a1/trcl_VA/flair/isoforms.productivity.psl'\n",
    "f=open(outfile)\n",
    "reader=csv.reader(f,delimiter='\\t')\n",
    "count=0\n",
    "for row in reader:\n",
    "    if int(row[-1])==2 and int(row[0])>=300: \n",
    "        count+=1\n",
    "        if int(row[0])>longest: longest=int(row[0])\n",
    "print count\n",
    "print 'longest single exon: '+str(longest)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Bedtools isoform-gene annotation using bedtools intersect function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Bedtools(directory,\n",
    "             isoforms,     #can be a gtf or psl file\n",
    "            annot,\n",
    "            outfile):      #GTF file\n",
    "    outfile=directory+'/'+outfile\n",
    "    infile=directory+'/'+isoforms\n",
    "    \n",
    "    if 'psl' in isoforms:\n",
    "        name=isoforms.split('.')[0]+'.gtf'\n",
    "        !python software/flair/bin/psl_to_gtf.py {infile} > {directory}/{name}\n",
    "        infile=directory+\"/\"+name\n",
    "    \n",
    "    !bedtools intersect -a {infile} -b {annot} -wb -s -split  > {outfile}\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    read_genes=[]\n",
    "    transcript_dict={}\n",
    "    overlap_transcripts=[]\n",
    "    \n",
    "    \n",
    "    for row in reader:\n",
    "        name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "        if name not in read_genes: read_genes.append(name)\n",
    "            \n",
    "        t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        if t not in transcript_dict: transcript_dict[t]=name\n",
    "        if name not in transcript_dict[t] and name not in overlap_transcripts:\n",
    "            overlap_transcripts.append(t)   \n",
    "    \n",
    "    f.seek(0)\n",
    "    \n",
    "    overlap_hits={}\n",
    "    for row in reader:\n",
    "        t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "        if t in overlap_transcripts:\n",
    "            if t in overlap_hits:\n",
    "                overlap_hits[t].append(name)\n",
    "            else:\n",
    "                overlap_hits[t]=[]\n",
    "                overlap_hits[t].append(name)\n",
    "    \n",
    "    for key in overlap_hits.keys():\n",
    "        del transcript_dict[key]\n",
    "        counts={}\n",
    "        for gene in overlap_hits[key]:\n",
    "            if gene in counts: counts[gene]+=1\n",
    "            else: counts[gene]=1\n",
    "        bestscore=0\n",
    "        bestgene=''\n",
    "        for key in counts:\n",
    "            if counts[key]>bestscore:\n",
    "                bestgene=key\n",
    "                bestscore=counts[key]\n",
    "        transcript_dict[t]=bestgene\n",
    "    \n",
    "    print 'Total genes expressed in '+directory+': '+str(len(read_genes))\n",
    "    print 'Total transcripts: ' +str(len(transcript_dict))\n",
    "    print str(len(overlap_transcripts))+' transcripts overlap multiple genes.'\n",
    "    \n",
    "    #Create isoform counts for every gene\n",
    "    gene_counts={}\n",
    "    for t in transcript_dict.keys():\n",
    "        if transcript_dict[t] not in gene_counts: gene_counts[transcript_dict[t]]=1\n",
    "        else: gene_counts[transcript_dict[t]]+=1\n",
    "    \n",
    "    #Create gene list\n",
    "    import pandas as pd\n",
    "    data=pd.DataFrame(list(gene_counts.items()))\n",
    "    export_path=\"/scratch/tshishido/\"+directory+\"/isosproductive.gene_list.csv\"\n",
    "    export_csv=data.to_csv(export_path,index=None,header=True)\n",
    "\n",
    "    print directory+' CHECK'\n",
    "    print 'After correcting overlaps:'\n",
    "    print 'Total transcripts: ' +str(len(transcript_dict))\n",
    "    print 'Number of genes from transcript count: '+ str(len(gene_counts))\n",
    "    \n",
    "    counts=[]\n",
    "    for gene in gene_counts.keys():\n",
    "        counts.append(gene_counts[gene])\n",
    "    largest=0\n",
    "    for num in counts:\n",
    "        if num>largest: largest=num\n",
    "    plt.hist(counts,edgecolor='black',bins=20,range=[1,largest])\n",
    "    plt.title('Isoform Distribution, '+directory)\n",
    "    plt.xlabel('Number of Isoforms per Gene')\n",
    "    plt.show()\n",
    "    count_counts={}\n",
    "    for num in counts:\n",
    "        if num in count_counts: count_counts[num]+=1\n",
    "        else: count_counts[num]=1\n",
    "    print count_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** WARNING: File software/flair/exonsgenes.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\texon\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; transcript_id \"ENST00000387314.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; transcript_type \"Mt_tRNA\"; transcript_name \"MT-TF-201\"; exon_number 1; exon_id \"ENSE00001544501.1\"; level 3; transcript_support_level \"NA\"; tag \"basic\";\n",
      "\n",
      "***** WARNING: File software/flair/exonsgenes.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\texon\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; transcript_id \"ENST00000387314.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; transcript_type \"Mt_tRNA\"; transcript_name \"MT-TF-201\"; exon_number 1; exon_id \"ENSE00001544501.1\"; level 3; transcript_support_level \"NA\"; tag \"basic\";\n",
      "\n",
      "Total genes expressed in 2019-7-26_cdna/a431.b2/trcl_VA/nc.flair: 7477\n",
      "Total transcripts: 10182\n",
      "22239 transcripts overlap multiple genes.\n",
      "2019-7-26_cdna/a431.b2/trcl_VA/nc.flair CHECK\n",
      "After correcting overlaps:\n",
      "Total transcripts: 10182\n",
      "Number of genes from transcript count: 6083\n"
     ]
    }
   ],
   "source": [
    "Bedtools('2019-7-26_cdna/a431.b2/trcl_VA/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/exonsgenes.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** WARNING: File software/flair/exonsgenes.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\texon\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; transcript_id \"ENST00000387314.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; transcript_type \"Mt_tRNA\"; transcript_name \"MT-TF-201\"; exon_number 1; exon_id \"ENSE00001544501.1\"; level 3; transcript_support_level \"NA\"; tag \"basic\";\n",
      "\n",
      "***** WARNING: File software/flair/exonsgenes.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\texon\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; transcript_id \"ENST00000387314.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; transcript_type \"Mt_tRNA\"; transcript_name \"MT-TF-201\"; exon_number 1; exon_id \"ENSE00001544501.1\"; level 3; transcript_support_level \"NA\"; tag \"basic\";\n",
      "\n",
      "Total genes expressed in 2019-7-26_cdna/a431.a1/trcl_VA/nc.flair: 8520\n",
      "Total transcripts: 13275\n",
      "27375 transcripts overlap multiple genes.\n",
      "2019-7-26_cdna/a431.a1/trcl_VA/nc.flair CHECK\n",
      "After correcting overlaps:\n",
      "Total transcripts: 13275\n",
      "Number of genes from transcript count: 6991\n"
     ]
    }
   ],
   "source": [
    "Bedtools('2019-7-26_cdna/a431.a1/trcl_VA/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/exonsgenes.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***** WARNING: File software/flair/new.gencode.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\tgene\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; level 3;\n",
      "\n",
      "***** WARNING: File software/flair/new.gencode.v29.annotation.gtf has inconsistent naming convention for record:\n",
      "chrM\tENSEMBL\tgene\t577\t647\t.\t+\t.\tgene_id \"ENSG00000210049.1\"; gene_type \"Mt_tRNA\"; gene_name \"MT-TF\"; level 3;\n",
      "\n",
      "Total genes expressed in 2019-7-26_cdna/hpmouse1/nc.flair: 145\n",
      "Total transcripts: 339\n",
      "339 transcripts overlap multiple genes.\n",
      "2019-7-26_cdna/hpmouse1/nc.flair CHECK\n",
      "After correcting overlaps:\n",
      "Total transcripts: 339\n",
      "Number of genes from transcript count: 135\n",
      "{1: 85, 2: 18, 3: 9, 4: 4, 5: 5, 6: 5, 7: 1, 9: 1, 10: 1, 11: 2, 14: 1, 17: 1, 20: 1, 21: 1}\n"
     ]
    }
   ],
   "source": [
    "Bedtools('2019-7-26_cdna/hpmouse1/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The number of transcripts accounted for in the output BAM generated by bedtools intersect is consistently smaller than the number of "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#debug code for dealing with isoforms intersecting with multiple genes\n",
    "\n",
    "outfile='2019-7-8_cdna/A431/trcl_VA/flair/isosproductive.annotated'\n",
    "reader=csv.reader(open(outfile),delimiter='\\t')\n",
    "transcript_dict={}\n",
    "\n",
    "for row in reader:\n",
    "    t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "    name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "    \n",
    "    if t not in transcript_dict:\n",
    "        transcript_dict[t]=[]\n",
    "        transcript_dict[t].append(name)\n",
    "    if t in transcript_dict:\n",
    "        if name not in transcript_dict[t]: transcript_dict[t].append(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b0428187-4017-4336-bc0a-92587948592f_NC\n",
      "{'CROCCP2': 6}\n",
      "b0428187-4017-4336-bc0a-92587948592f_NC, best gene: CROCCP2 6\n",
      "b243ea2c-f568-4cd4-820c-c88cc6a9e6e4\n",
      "{'PPIE': 93}\n",
      "b243ea2c-f568-4cd4-820c-c88cc6a9e6e4, best gene: PPIE 93\n",
      "8a25c52d-a31f-4f3f-a898-afdd6bc9b903\n",
      "{'MTF2': 100}\n",
      "8a25c52d-a31f-4f3f-a898-afdd6bc9b903, best gene: MTF2 100\n",
      "0c8ceeaf-b971-4aec-9cb1-c42553b77d75_NC\n",
      "{'RBBP4': 2}\n",
      "0c8ceeaf-b971-4aec-9cb1-c42553b77d75_NC, best gene: RBBP4 2\n",
      "62dc2f3b-13e8-4c71-b26e-2c66480d2f39_NC\n",
      "{'ATF6': 2}\n",
      "62dc2f3b-13e8-4c71-b26e-2c66480d2f39_NC, best gene: ATF6 2\n",
      "10f56ffc-ffc0-4d54-bf61-801a35569424_NC\n",
      "{'PKN2': 2}\n",
      "10f56ffc-ffc0-4d54-bf61-801a35569424_NC, best gene: PKN2 2\n",
      "bd8f3e68-4560-42df-aac3-cbffde29181c\n",
      "{'PTPRU': 133}\n",
      "bd8f3e68-4560-42df-aac3-cbffde29181c, best gene: PTPRU 133\n",
      "f99df470-176f-4317-938a-2602a619a46d\n",
      "{'LINC00623': 5}\n",
      "f99df470-176f-4317-938a-2602a619a46d, best gene: LINC00623 5\n",
      "f77e34b7-c5fa-467f-8f21-011a248e8eda_NC\n",
      "{'DCAF6': 6}\n",
      "f77e34b7-c5fa-467f-8f21-011a248e8eda_NC, best gene: DCAF6 6\n",
      "d3459c1b-d392-4a7a-bc4b-f5d90cbf1826_NC\n",
      "{'UCK2': 9, 'MIR3658': 2}\n",
      "d3459c1b-d392-4a7a-bc4b-f5d90cbf1826_NC, best gene: UCK2 9\n",
      "ENST00000263702.10\n",
      "{'MECR': 107}\n",
      "ENST00000263702.10, best gene: MECR 107\n",
      "e546ab9d-a44f-4ca1-8272-a8c46c0d5ec7\n",
      "{'LINC01133': 6}\n",
      "e546ab9d-a44f-4ca1-8272-a8c46c0d5ec7, best gene: LINC01133 6\n",
      "97c6222c-bcd8-4a61-ab8b-34c477af6ac0_NC\n",
      "{'CELSR2': 3}\n",
      "97c6222c-bcd8-4a61-ab8b-34c477af6ac0_NC, best gene: CELSR2 3\n",
      "ENST00000466803.1\n",
      "{'HOOK1': 22}\n",
      "ENST00000466803.1, best gene: HOOK1 22\n",
      "59ae1ce7-4230-46ef-9ab0-3233c7e8f11b_NC\n",
      "{'PIGC': 3}\n",
      "59ae1ce7-4230-46ef-9ab0-3233c7e8f11b_NC, best gene: PIGC 3\n",
      "ENST00000370550.9\n",
      "{'AC093155.3': 11, 'HS2ST1': 22}\n",
      "ENST00000370550.9, best gene: HS2ST1 22\n",
      "34613a37-df3c-4af8-942b-b6ffe2581fb2\n",
      "{'TTC4': 31, 'MROH7-TTC4': 46}\n",
      "34613a37-df3c-4af8-942b-b6ffe2581fb2, best gene: MROH7-TTC4 46\n",
      "2b05c92e-f896-4128-bed5-d2d4d59d2407_NC\n",
      "{'AC092803.2': 2}\n",
      "2b05c92e-f896-4128-bed5-d2d4d59d2407_NC, best gene: AC092803.2 2\n",
      "a9de9c58-9906-46f5-bd3c-bfd297fdfc22_NC\n",
      "{'ATP2B4': 4}\n",
      "a9de9c58-9906-46f5-bd3c-bfd297fdfc22_NC, best gene: ATP2B4 4\n",
      "ENST00000338513.1\n",
      "{'GJB5': 4}\n",
      "ENST00000338513.1, best gene: GJB5 4\n",
      "27f5b6e3-69d6-47a3-a017-8cf6bffbb915\n",
      "{'LAD1': 29}\n",
      "27f5b6e3-69d6-47a3-a017-8cf6bffbb915, best gene: LAD1 29\n",
      "6a906360-bcf6-40d3-8cd8-4b2548e87436\n",
      "{'DHDDS': 96}\n",
      "6a906360-bcf6-40d3-8cd8-4b2548e87436, best gene: DHDDS 96\n",
      "a16a22a2-c154-473f-8971-b136c7fdf116_NC\n",
      "{'NCDN': 4}\n",
      "a16a22a2-c154-473f-8971-b136c7fdf116_NC, best gene: NCDN 4\n",
      "407de77a-79a6-4211-87e8-cd137ef7d862_NC\n",
      "{'IFI6': 4}\n",
      "407de77a-79a6-4211-87e8-cd137ef7d862_NC, best gene: IFI6 4\n",
      "38f4161c-dd25-46c2-bf82-d560a20c77b5\n",
      "{'CAP1': 169}\n",
      "38f4161c-dd25-46c2-bf82-d560a20c77b5, best gene: CAP1 169\n",
      "122da8d1-c05c-4cad-9150-2d329b3e1e43\n",
      "{'ANP32E': 29}\n",
      "122da8d1-c05c-4cad-9150-2d329b3e1e43, best gene: ANP32E 29\n",
      "8f238fbf-6d74-4775-ae80-effb84cb88af\n",
      "{'RPL7P9': 2}\n",
      "8f238fbf-6d74-4775-ae80-effb84cb88af, best gene: RPL7P9 2\n",
      "4d33d8aa-1739-47ba-8582-0c5784f9121b_NC\n",
      "{'EMC1': 2}\n",
      "4d33d8aa-1739-47ba-8582-0c5784f9121b_NC, best gene: EMC1 2\n",
      "5705a076-a391-4486-a772-b8a521782713\n",
      "{'IVNS1ABP': 71}\n",
      "5705a076-a391-4486-a772-b8a521782713, best gene: IVNS1ABP 71\n",
      "a38d6321-393c-466e-8793-a14b926bac08\n",
      "{'USP48': 39}\n",
      "a38d6321-393c-466e-8793-a14b926bac08, best gene: USP48 39\n",
      "de66a150-4a68-4a5c-bc86-e6637557020e_NC\n",
      "{'KDM1A': 6}\n",
      "de66a150-4a68-4a5c-bc86-e6637557020e_NC, best gene: KDM1A 6\n",
      "52c30eca-204c-4404-8287-35c931ea1ae7_NC\n",
      "{'ZMYM1': 6}\n",
      "52c30eca-204c-4404-8287-35c931ea1ae7_NC, best gene: ZMYM1 6\n",
      "2b7ff30b-338d-4c2b-8e41-5eb0a75a42c8_NC\n",
      "{'PRPF38B': 3}\n",
      "2b7ff30b-338d-4c2b-8e41-5eb0a75a42c8_NC, best gene: PRPF38B 3\n",
      "0ed4ed1e-0b5e-444f-90c3-ef433822c7eb_NC\n",
      "{'KCNJ10': 1, 'PIGM': 2}\n",
      "0ed4ed1e-0b5e-444f-90c3-ef433822c7eb_NC, best gene: PIGM 2\n",
      "3d5a967d-c478-431f-9b1e-26245e43f845_NC\n",
      "{'LAMC1': 2}\n",
      "3d5a967d-c478-431f-9b1e-26245e43f845_NC, best gene: LAMC1 2\n",
      "ENST00000562225.1\n",
      "{'IFI16': 40}\n",
      "ENST00000562225.1, best gene: IFI16 40\n",
      "dafc6ba9-2c99-4a31-a7e5-bac594789281\n",
      "{'VPS13D': 12}\n",
      "dafc6ba9-2c99-4a31-a7e5-bac594789281, best gene: VPS13D 12\n",
      "5916ffcf-2c71-4916-a277-3224f9f341c2_NC\n",
      "{'PPP1R15B': 2}\n",
      "5916ffcf-2c71-4916-a277-3224f9f341c2_NC, best gene: PPP1R15B 2\n",
      "6cce91db-2cc8-4200-bb96-9a0037787b37_NC\n",
      "{'RAB7B': 4}\n",
      "6cce91db-2cc8-4200-bb96-9a0037787b37_NC, best gene: RAB7B 4\n",
      "f93d9f43-b084-4c0f-8564-88e768bc5b72\n",
      "{'TMEM54': 34}\n",
      "f93d9f43-b084-4c0f-8564-88e768bc5b72, best gene: TMEM54 34\n",
      "cba18d90-3e39-4634-a3b2-5ce937d2ec09\n",
      "{'MAPKAPK2': 20}\n",
      "cba18d90-3e39-4634-a3b2-5ce937d2ec09, best gene: MAPKAPK2 20\n",
      "88929cd5-f010-4d9f-b4f6-99d855297fa4\n",
      "{'GNL2': 24}\n",
      "88929cd5-f010-4d9f-b4f6-99d855297fa4, best gene: GNL2 24\n",
      "d96d3561-af2a-4565-a979-55fbf5c6aef4_NC\n",
      "{'DDX20': 3}\n",
      "d96d3561-af2a-4565-a979-55fbf5c6aef4_NC, best gene: DDX20 3\n",
      "ENST00000339083.11\n",
      "{'RHOC': 96, 'AL603832.3': 21}\n",
      "ENST00000339083.11, best gene: RHOC 96\n",
      "36fc55b3-a092-4f7e-a6cc-32d37f442912\n",
      "{'SNHG28': 7, 'CFAP45': 15}\n",
      "36fc55b3-a092-4f7e-a6cc-32d37f442912, best gene: CFAP45 15\n",
      "ENST00000340006.6\n",
      "{'CSRP1': 89}\n",
      "ENST00000340006.6, best gene: CSRP1 89\n",
      "11d3e1af-f9bb-4c5b-8724-295640e35aa0_NC\n",
      "{'CRYZL2P': 2, 'CRYZL2P-SEC16B': 1}\n",
      "11d3e1af-f9bb-4c5b-8724-295640e35aa0_NC, best gene: CRYZL2P 2\n",
      "ENST00000366621.7\n",
      "{'KCNK1': 14}\n",
      "ENST00000366621.7, best gene: KCNK1 14\n",
      "ENST00000496025.1\n",
      "{'TMEM63A': 7}\n",
      "ENST00000496025.1, best gene: TMEM63A 7\n",
      "ee0c6737-f94d-4c48-81eb-c1b25b671373_NC\n",
      "{'TOR1AIP2': 2}\n",
      "ee0c6737-f94d-4c48-81eb-c1b25b671373_NC, best gene: TOR1AIP2 2\n",
      "aaf25a82-ee55-4229-b1c8-6f07cf0b8219\n",
      "{'KCTD3': 19}\n",
      "aaf25a82-ee55-4229-b1c8-6f07cf0b8219, best gene: KCTD3 19\n",
      "cd55fedb-c694-4a64-b1ef-535aa1d19b2c\n",
      "{'HEATR1': 12}\n",
      "cd55fedb-c694-4a64-b1ef-535aa1d19b2c, best gene: HEATR1 12\n",
      "ENST00000369109.7\n",
      "{'APH1A': 46}\n",
      "ENST00000369109.7, best gene: APH1A 46\n",
      "1b17989a-00c7-4d63-8cfb-018211f5bc3b\n",
      "{'COPA': 157}\n",
      "1b17989a-00c7-4d63-8cfb-018211f5bc3b, best gene: COPA 157\n",
      "85fc8279-ca95-4640-a968-33578c751984_NC\n",
      "{'PHGDH': 17}\n",
      "85fc8279-ca95-4640-a968-33578c751984_NC, best gene: PHGDH 17\n",
      "211e639b-63e5-456a-8cc6-7912958fa8db-1_NC\n",
      "{'ENAH': 3}\n",
      "211e639b-63e5-456a-8cc6-7912958fa8db-1_NC, best gene: ENAH 3\n",
      "8049dce9-7658-4e47-b6cd-b621e2b7bdc5_NC\n",
      "{'WDTC1': 6}\n",
      "8049dce9-7658-4e47-b6cd-b621e2b7bdc5_NC, best gene: WDTC1 6\n",
      "1e1fe701-146f-4d0e-9ed0-300be7771f49_NC\n",
      "{'MAN1A2': 2}\n",
      "1e1fe701-146f-4d0e-9ed0-300be7771f49_NC, best gene: MAN1A2 2\n",
      "2f1605af-3807-43e3-916e-6c18005dce1e_NC\n",
      "{'MIER1': 8}\n",
      "2f1605af-3807-43e3-916e-6c18005dce1e_NC, best gene: MIER1 8\n",
      "ENST00000361531.6\n",
      "{'MRPL24': 32}\n",
      "ENST00000361531.6, best gene: MRPL24 32\n",
      "a84c9626-3858-4d4b-a380-41547ac76650_NC\n",
      "{'AGTRAP': 27}\n",
      "a84c9626-3858-4d4b-a380-41547ac76650_NC, best gene: AGTRAP 27\n",
      "71d968d5-0142-450a-adad-95db3af76208_NC\n",
      "{'NEK7': 4}\n",
      "71d968d5-0142-450a-adad-95db3af76208_NC, best gene: NEK7 4\n",
      "fb696087-de90-4aa6-a3ec-54092bffb7b7_NC\n",
      "{'MTR': 5}\n",
      "fb696087-de90-4aa6-a3ec-54092bffb7b7_NC, best gene: MTR 5\n",
      "e791f709-201f-49d6-9ef0-ea999093326d_NC\n",
      "{'RPL21P28': 2}\n",
      "e791f709-201f-49d6-9ef0-ea999093326d_NC, best gene: RPL21P28 2\n",
      "40f10297-a139-4904-91d7-f8b59348e681\n",
      "{'ODF2L': 42}\n",
      "40f10297-a139-4904-91d7-f8b59348e681, best gene: ODF2L 42\n",
      "e7272271-ab0b-4b93-9d57-33c798a824bf\n",
      "{'TTF2': 22}\n",
      "e7272271-ab0b-4b93-9d57-33c798a824bf, best gene: TTF2 22\n",
      "19ff474a-578a-4942-8922-31ea5e94c253_NC\n",
      "{'METTL18': 4}\n",
      "19ff474a-578a-4942-8922-31ea5e94c253_NC, best gene: METTL18 4\n",
      "b3913dd7-2cd0-485b-abc7-2503352f6921_NC\n",
      "{'SNAPIN': 5}\n",
      "b3913dd7-2cd0-485b-abc7-2503352f6921_NC, best gene: SNAPIN 5\n",
      "ENST00000361532.9\n",
      "{'ENSA': 42}\n",
      "ENST00000361532.9, best gene: ENSA 42\n",
      "2c391d34-76dd-411c-a668-67e445b2f4fc_NC\n",
      "{'LINC02474': 2}\n",
      "2c391d34-76dd-411c-a668-67e445b2f4fc_NC, best gene: LINC02474 2\n",
      "ENST00000455785.6\n",
      "{'MIR3917': 2, 'STMN1': 36}\n",
      "ENST00000455785.6, best gene: STMN1 36\n",
      "77fbfe82-564a-4f79-8181-fc3e2ff34c4d_NC\n",
      "{'B3GALT6': 3}\n",
      "77fbfe82-564a-4f79-8181-fc3e2ff34c4d_NC, best gene: B3GALT6 3\n",
      "6f4d47cc-16d3-4490-affd-8e4cc9c27624_NC\n",
      "{'TUFT1': 5}\n",
      "6f4d47cc-16d3-4490-affd-8e4cc9c27624_NC, best gene: TUFT1 5\n",
      "ENST00000366628.9\n",
      "{'NTPCR': 33}\n",
      "ENST00000366628.9, best gene: NTPCR 33\n",
      "af8869ab-ae61-4d21-9b2b-694798e261f2\n",
      "{'TIMM17A': 16}\n",
      "af8869ab-ae61-4d21-9b2b-694798e261f2, best gene: TIMM17A 16\n",
      "89b5b7f5-4261-4119-8253-1a62dbc3aa7a_NC\n",
      "{'C1orf159': 10}\n",
      "89b5b7f5-4261-4119-8253-1a62dbc3aa7a_NC, best gene: C1orf159 10\n",
      "16128abb-6a06-4b4a-9014-35a74c7d1bc7_NC\n",
      "{'ABCB10': 2}\n",
      "16128abb-6a06-4b4a-9014-35a74c7d1bc7_NC, best gene: ABCB10 2\n",
      "ENST00000372969.7\n",
      "{'NDUFS5': 9}\n",
      "ENST00000372969.7, best gene: NDUFS5 9\n",
      "bba14584-6931-4731-9bb8-388eba0cc6f7_NC\n",
      "{'IVL': 2}\n",
      "bba14584-6931-4731-9bb8-388eba0cc6f7_NC, best gene: IVL 2\n",
      "69617591-766d-4451-9768-d4c004827638\n",
      "{'ACAP3': 25}\n",
      "69617591-766d-4451-9768-d4c004827638, best gene: ACAP3 25\n",
      "abd45c97-8f43-4ac1-a3e3-88e114d1fdbe_NC\n",
      "{'AL031428.1': 2, 'LUZP1': 3}\n",
      "abd45c97-8f43-4ac1-a3e3-88e114d1fdbe_NC, best gene: LUZP1 3\n",
      "781df769-f2b6-47dd-bca8-fc1458712a07\n",
      "{'EPHX4': 13}\n",
      "781df769-f2b6-47dd-bca8-fc1458712a07, best gene: EPHX4 13\n",
      "ENST00000368376.7\n",
      "{'MTX1': 56}\n",
      "ENST00000368376.7, best gene: MTX1 56\n",
      "9feca0cc-871a-4775-a5cf-0e069a35ad09_NC\n",
      "{'C1orf74': 2}\n",
      "9feca0cc-871a-4775-a5cf-0e069a35ad09_NC, best gene: C1orf74 2\n",
      "226273c9-a67c-4e60-819a-ae85a510529a_NC\n",
      "{'IPO9': 2}\n",
      "226273c9-a67c-4e60-819a-ae85a510529a_NC, best gene: IPO9 2\n",
      "8436f94f-fda3-49ee-973e-de6b5f041856_NC\n",
      "{'SARS': 4}\n",
      "8436f94f-fda3-49ee-973e-de6b5f041856_NC, best gene: SARS 4\n",
      "0e06aef0-76a0-487c-a0a5-734d90f5b240_NC\n",
      "{'BTF3L4': 3}\n",
      "0e06aef0-76a0-487c-a0a5-734d90f5b240_NC, best gene: BTF3L4 3\n",
      "2e4c85d5-f99f-44ac-bda7-ff23f079ad9a\n",
      "{'NME7': 20}\n",
      "2e4c85d5-f99f-44ac-bda7-ff23f079ad9a, best gene: NME7 20\n",
      "8074c8c4-5ed8-4243-be48-b8547cf89671\n",
      "{'USP48': 34}\n",
      "8074c8c4-5ed8-4243-be48-b8547cf89671, best gene: USP48 34\n",
      "c302eaa5-bd6e-427e-9d4f-4a7775e384de_NC\n",
      "{'TRMT13': 2}\n",
      "c302eaa5-bd6e-427e-9d4f-4a7775e384de_NC, best gene: TRMT13 2\n",
      "9fd6c4d0-45a5-4109-9d66-cbc6b61072d8\n",
      "{'MIIP': 31}\n",
      "9fd6c4d0-45a5-4109-9d66-cbc6b61072d8, best gene: MIIP 31\n",
      "ENST00000334047.11\n",
      "{'F3': 23}\n",
      "ENST00000334047.11, best gene: F3 23\n",
      "741082d7-eb11-4ce2-8bd5-1121cf9a08ea_NC\n",
      "{'CTTNBP2NL': 2}\n",
      "741082d7-eb11-4ce2-8bd5-1121cf9a08ea_NC, best gene: CTTNBP2NL 2\n",
      "ENST00000489308.6\n",
      "{'BTF3L4': 38}\n",
      "ENST00000489308.6, best gene: BTF3L4 38\n",
      "78603216-df78-447e-a7a4-089784d6e33f-1_NC\n",
      "{'TBCE': 7, 'GGPS1': 7}\n",
      "78603216-df78-447e-a7a4-089784d6e33f-1_NC, best gene: TBCE 7\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'TBCE'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-590-17fadd6cfe0b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m', best gene: '\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mbestgene\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m' '\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbestscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m     \u001b[0mgene_counts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbestgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'TBCE'"
     ]
    }
   ],
   "source": [
    "# attempt to debug transcripts intersecting multiple genes\n",
    "\n",
    "file=\"2019-6-25_dirrna/trcl_VA/flair/isoforms.annotated\"\n",
    "f=open(file)\n",
    "        \n",
    "for x in overlap_transcripts:\n",
    "    print x\n",
    "    counts={}\n",
    "    f.seek(0)     ##Resets reader to line 0\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    for row in reader:\n",
    "        transcript=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        gene=row[17].split('gene_name \"')[1].split('\"')[0]\n",
    "        if transcript in x:\n",
    "            if gene in counts:\n",
    "                counts[gene]+=1\n",
    "            else:\n",
    "                counts[gene]=1\n",
    "    print counts\n",
    "        \n",
    "    bestscore=0\n",
    "    bestgene=''\n",
    "    for gene in counts:\n",
    "        if counts[gene]>bestscore:\n",
    "            bestgene=gene\n",
    "            bestscore=counts[gene]\n",
    "    print x+', best gene: '+bestgene+' '+str(bestscore)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Other notes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Other notes\n",
    "\n",
    "#Created fasta reference genome for only human chr.21\n",
    "!sed -n \"/^>NC_000021.9$/,/^>/p\" software/homo_GRCh38_trimmed_ref.fa >  software/homo_chr21.GRCh38_trimmed_ref.fa\n",
    "\n",
    "#Created human GENCODE annotation file of only exon and gene entries, to use for bedtools intersect\n",
    "!grep '\tgene\t' software/flair/new.gencode.v29.annotation.gtf > software/flair/genes.v29.annotation.gtf \n",
    "!grep '\texon\t' software/flair/new.gencode.v29.annotation.gtf > software/flair/exons.v29.annotation.gtf\n",
    "!cat software/flair/exons.v29.annotation.gtf software/flair/genes.v29.annotation.gtf > software/flair/exonsgenes.v29.annotation.gtf\n",
    "\n",
    "\n",
    "#Did not get around to using these 2 flair scripts but could be of good use for \n",
    "    ##identifying differential isoform expression\n",
    "    ##1)\n",
    "count_file='2019-7-26_cdna/a431.a1/trcl_VA/flair/counts_matrix.tsv'\n",
    "out_dir='2019-7-26_cdna/a431.a1/trcl_VA/flair/diffexp'\n",
    "!python software/flair/flair.py diffExp -q {count_file} -o {out_dir}\n",
    "    ##2)\n",
    "#python diff_iso_usage.py isoforms.psl colnum1 colnum2 diff_isos.txt\n",
    "isoforms='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/3.flair.collapse.isoforms.psl' \n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/diff_isos.txt'\n",
    "!python software/flair/bin/diff_iso_usage.py {isoforms} {outfile}\n",
    "\n",
    "#Bedtools annotate function, did not look into\n",
    "annot='software/flair/exonsgenes.v29.annotation.gtf'\n",
    "isoforms='2019-7-8_cdna/A431/trcl_VA/flair/3.flair.collapse.isoforms.gtf'\n",
    "output=\"2019-7-8_cdna/A431/trcl_VA/flair/bedtools.annotate\"\n",
    "!bedtools annotate -i {annot} -files {isoforms} > {output}\n",
    "\n",
    "#This flair script didn't do anything, just reappended on the chromosome position\n",
    "isoforms='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.collapse.isoforms.psl' \n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.named'\n",
    "annot='software/flair/new.gencode.v29.annotation.gtf'\n",
    "!python software/flair/bin/identify_gene_isoform.py {isoforms} {annot} {outfile}\n",
    "\n",
    "#Flair script identifies gene annotation for each entry of PSL. Only corrected a fraction of the isoforms\n",
    "    ##missing genes annotations\n",
    "\n",
    "psl='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.collapse.isoforms.psl'\n",
    "ref='software/flair/new.gencode.v29.annotation.gtf'\n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/isos_matched.psl'\n",
    "!python software/flair/bin/identify_annotated_gene.py {psl} {ref} {outfile}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
