{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!git clone https://github.com/BrooksLabUCSC/flair.git\n",
    "!pip install kerneltree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import csv\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Script for Flair. 1) Converts bam mapped reads file to bed12 and corrects misaligned splice sites using a genome annotation. 2) Collapse corrected reads into high-confidence isoforms and quantifies them by counting converted into tpm.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Flair(\n",
    "    main_dir, #full path of working directory\n",
    "    bam_input, #only filename, path same as main_dir\n",
    "    genome,    #full path of reference genome (in ref_fasta folder)\n",
    "    chr_sizes, #full path to tab-delim file of chromosome sizes, names must correspond to both annotation and reference\n",
    "    chr_annot, #full path to gtf annotation file\n",
    "    reads,     #full path to fa file of UN-mapped raw reads\n",
    "    file_org): #full path to tab-delim file listing sample information\n",
    "    \n",
    "    !mkdir {main_dir}/nc.flair\n",
    "#     !samtools faidx {genome}  <- if using a new genome fasta file\n",
    "    print(\"Copy paste into command line window\")\n",
    "    print(\"1 convert, correct, collapse scripts:\") ##**NB: module tqdm only installed on python3, not just python\n",
    "    #both samtools and minimap2 directories are located in software/flair/\n",
    "        #-s parameter is minimum number of supporting reads for an isoform (default: 3) \n",
    "        #--stringent\n",
    "    print('$python3 software/flair/bin/bam2Bed12.py -i '+main_dir+'/'+bam_input+' > '+ main_dir+'/nc.flair/_clean_sorted.bed12;python3 software/flair/flair.py correct -c '+chr_sizes+' -g '+genome+' -q '+main_dir+'/nc.flair/_clean_sorted.bed12 -o '+main_dir+'/nc.flair/flair -f '+chr_annot+'; time python3 software/flair/flair.py collapse -g '+genome+' -r '+reads+' -f '+chr_annot+' -o '+main_dir+\"/nc.flair/flair.collapse -q \"+main_dir+'/nc.flair/flair_all_corrected.psl')\n",
    "    main_dir=main_dir+'/nc.flair'\n",
    "    print \"2 quantify:\"\n",
    "    print 'python3 software/flair/flair.py quantify --tpm -r '+file_org+' -o '+main_dir+'/counts_matrix.tsv -i '+main_dir+'/flair.collapse.isoforms.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Flair(\n",
    "    main_dir='2019-8-6_cdna/a431.b1.merged',\n",
    "    bam_input='_cleansort.bam',\n",
    "    genome='software/ref_fasta/homo_GRCh38_trimmed_ref.fa',\n",
    "    chr_sizes='software/bedtools2/genomes/clean.human.hg38.genome.chrom.sizes.tsv',\n",
    "    chr_annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "    reads='2019-8-6_cdna/a431.b1/a431.b1.fasta',   ##use raw, unaligned reads\n",
    "    file_org='2019-7-26_cdna.txt'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Isoform productivity: append label to end of each line of PSL\n",
    "    ##0: functional stop codon (productive)\n",
    "    ##1: premature stop codon (unproductive)\n",
    "    ##2: lncRNA (no start codon) \n",
    "    \n",
    "    #From script: 'Unproductive proportion estimate: unproductive / (valid_transcripts + unproductive))\n",
    "    ##Single-exon transcripts are counted as lncRNAs.\n",
    "\n",
    "def IsoformProductivity(directory,\n",
    "                       reads,\n",
    "                       annot,\n",
    "                       genome):\n",
    "    reads=directory+reads\n",
    "    outfile=directory+'/isoforms.productivity.psl'\n",
    "#     outfile=directory+'/isosproductive.psl'\n",
    "    !python software/flair/bin/mark_productivity.py {reads} {annot} {genome} > {outfile}\n",
    "\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "\n",
    "    good_stop=0\n",
    "    bad_stop=0\n",
    "    no_start=0\n",
    "    single_exon=0\n",
    "    for row in reader:\n",
    "        n=int(row[-1])\n",
    "        if n==0: good_stop+=1\n",
    "        elif n==1: bad_stop+=1\n",
    "        elif n==2: no_start+=1\n",
    "        if int(row[17])==1: single_exon+=1\n",
    "\n",
    "    print directory        \n",
    "    print 'Productive isoforms: '+str(good_stop)\n",
    "    print 'Unproductive isoforms: '+str(bad_stop)\n",
    "    total=good_stop+bad_stop+no_start\n",
    "    print 'Proportion of productive: ' +str(good_stop*100 / (total))+'%'\n",
    "    print 'lncRNA/other: '+str(no_start - single_exon)\n",
    "    print 'Single exon transcripts: '+str(single_exon)\n",
    "    Productive_isos(directory,outfile)\n",
    "\n",
    "def Productive_isos(directory,\n",
    "                   infile):\n",
    "    out=directory+'/all_productive_isos.psl'\n",
    "    i=open(infile,'r')\n",
    "    o=open(out,'a')\n",
    "    reader=csv.reader(i,delimiter='\\t')\n",
    "    writer = csv.writer(o,delimiter='\\t')\n",
    "    for row in reader:\n",
    "        length=abs(int(row[16])-int(row[15]))\n",
    "        if int(row[-1])==2 and length>=300: writer.writerow(row)\n",
    "        elif int(row[21])==0: writer.writerow(row)   \n",
    "\n",
    "##Marks each isoform by appending new column to each entry of PSL\n",
    "    ##0: spliced\n",
    "    ##1: intron-retaining\n",
    "    \n",
    "def IntronRetention(directory,\n",
    "                   isoforms='exonsgenes.collapse.isoforms.psl',\n",
    "                   outfile='isoforms.ir.psl'):\n",
    "    isoforms=directory+'/'+isoforms \n",
    "    outfile=directory+'/'+outfile\n",
    "    coords=directory+'/coords.txt'\n",
    "    !python software/flair/bin/mark_intron_retention.py {isoforms} {outfile} {coords}\n",
    "\n",
    "    ##Count intron-retaining and spliced isoforms\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    spliced=0\n",
    "    intron_retain=0\n",
    "    splice_pos=0\n",
    "    splice_neg=0\n",
    "    retain_pos=0\n",
    "    retain_neg=0\n",
    "    for row in reader:\n",
    "        if int(row[21])==0: spliced+=1\n",
    "        elif int(row[21])==1: intron_retain+=1\n",
    "\n",
    "    #Count spliced isoforms based on strand\n",
    "        if int(row[21])==0 and row[8] is '+': splice_pos+=1\n",
    "        elif int(row[21])==0 and row[8] is '-': splice_neg+=1\n",
    "        if int(row[21])==1 and row[8] is '+': retain_pos+=1\n",
    "        elif int(row[21])==1 and row[8] is '-': retain_neg+=1    \n",
    "\n",
    "    print directory\n",
    "    print 'Spliced isoforms: '+str(spliced)\n",
    "    print 'Intron-retaining isoforms: '+str(intron_retain)\n",
    "    print '+ strand, spliced: '+str(splice_pos)\n",
    "    print '- strand, spliced: '+str(splice_neg)\n",
    "    print '+ strand, retaining: '+str(retain_pos)\n",
    "    print '+ strand, retaining: '+str(retain_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IsoformProductivity('2019-8-6_cdna/a431.b1/trcl_VA/nc.flair',\n",
    "                       reads='/flair.collapse.isoforms.psl',\n",
    "                       annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "                       genome='2019-6-13_A431/GRCh38_trimmed.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IntronRetention('2019-7-22_dirrna/trcl_VA/flair',\n",
    "                   isoforms='isosproductive.psl',\n",
    "                   outfile='isosproductive.ir.psl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#debug code to determine longest single exon transcript\n",
    "\n",
    "f=open('2019-7-26_cdna/a431utc/trcl_VA/flair/isoforms.productivity.psl')\n",
    "reader=csv.reader(f,delimiter='\\t')\n",
    "count=0\n",
    "longest=0\n",
    "printrow=''\n",
    "for row in reader:\n",
    "    if int(row[-1])==2 and int(row[0])>=300: \n",
    "        count+=1\n",
    "        if int(row[0])>longest: longest=int(row[0])\n",
    "\n",
    "print count\n",
    "print 'longest single exon: '+str(longest)\n",
    "f.close()\n",
    "        \n",
    "outfile='2019-7-26_cdna/a431.a1/trcl_VA/flair/isoforms.productivity.psl'\n",
    "f=open(outfile)\n",
    "reader=csv.reader(f,delimiter='\\t')\n",
    "count=0\n",
    "for row in reader:\n",
    "    if int(row[-1])==2 and int(row[0])>=300: \n",
    "        count+=1\n",
    "        if int(row[0])>longest: longest=int(row[0])\n",
    "print count\n",
    "print 'longest single exon: '+str(longest)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Bedtools isoform-gene annotation using bedtools intersect function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Bedtools(directory,\n",
    "             isoforms,     #can be a gtf or psl file\n",
    "            annot,\n",
    "            outfile):      #GTF file\n",
    "    outfile=directory+'/'+outfile\n",
    "    infile=directory+'/'+isoforms\n",
    "    \n",
    "    if 'psl' in isoforms:\n",
    "        name=isoforms.split('.')[0]+'.gtf'\n",
    "        !python software/flair/bin/psl_to_gtf.py {infile} > {directory}/{name}\n",
    "        infile=directory+\"/\"+name\n",
    "    \n",
    "    !bedtools intersect -a {infile} -b {annot} -wb -s -split  > {outfile}\n",
    "    f=open(outfile)\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    read_genes=[]\n",
    "    transcript_dict={}\n",
    "    overlap_transcripts=[]\n",
    "    \n",
    "    \n",
    "    for row in reader:\n",
    "        name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "        if name not in read_genes: read_genes.append(name)\n",
    "            \n",
    "        t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        if t not in transcript_dict: transcript_dict[t]=name\n",
    "        if name not in transcript_dict[t] and name not in overlap_transcripts:\n",
    "            overlap_transcripts.append(t)   \n",
    "    \n",
    "    f.seek(0)\n",
    "    \n",
    "    overlap_hits={}\n",
    "    for row in reader:\n",
    "        t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "        if t in overlap_transcripts:\n",
    "            if t in overlap_hits:\n",
    "                overlap_hits[t].append(name)\n",
    "            else:\n",
    "                overlap_hits[t]=[]\n",
    "                overlap_hits[t].append(name)\n",
    "    \n",
    "    for key in overlap_hits.keys():\n",
    "        del transcript_dict[key]\n",
    "        counts={}\n",
    "        for gene in overlap_hits[key]:\n",
    "            if gene in counts: counts[gene]+=1\n",
    "            else: counts[gene]=1\n",
    "        bestscore=0\n",
    "        bestgene=''\n",
    "        for key in counts:\n",
    "            if counts[key]>bestscore:\n",
    "                bestgene=key\n",
    "                bestscore=counts[key]\n",
    "        transcript_dict[t]=bestgene\n",
    "    \n",
    "    print 'Total genes expressed in '+directory+': '+str(len(read_genes))\n",
    "    print 'Total transcripts: ' +str(len(transcript_dict))\n",
    "    print str(len(overlap_transcripts))+' transcripts overlap multiple genes.'\n",
    "    \n",
    "    #Create isoform counts for every gene\n",
    "    gene_counts={}\n",
    "    for t in transcript_dict.keys():\n",
    "        if transcript_dict[t] not in gene_counts: gene_counts[transcript_dict[t]]=1\n",
    "        else: gene_counts[transcript_dict[t]]+=1\n",
    "    \n",
    "    #Create gene list\n",
    "    import pandas as pd\n",
    "    data=pd.DataFrame(list(gene_counts.items()))\n",
    "    export_path=\"/scratch/tshishido/\"+directory+\"/isosproductive.gene_list.csv\"\n",
    "    export_csv=data.to_csv(export_path,index=None,header=True)\n",
    "\n",
    "    print directory+' CHECK'\n",
    "    print 'After correcting overlaps:'\n",
    "    print 'Total transcripts: ' +str(len(transcript_dict))\n",
    "    print 'Number of genes from transcript count: '+ str(len(gene_counts))\n",
    "    \n",
    "    counts=[]\n",
    "    for gene in gene_counts.keys():\n",
    "        counts.append(gene_counts[gene])\n",
    "    largest=0\n",
    "    for num in counts:\n",
    "        if num>largest: largest=num\n",
    "    plt.hist(counts,edgecolor='black',bins=20,range=[1,largest])\n",
    "    plt.title('Isoform Distribution, '+directory)\n",
    "    plt.xlabel('Number of Isoforms per Gene')\n",
    "    plt.show()\n",
    "    count_counts={}\n",
    "    for num in counts:\n",
    "        if num in count_counts: count_counts[num]+=1\n",
    "        else: count_counts[num]=1\n",
    "    print count_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bedtools('2019-7-26_cdna/a431.b2/trcl_VA/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/exonsgenes.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bedtools('2019-7-26_cdna/a431.a1/trcl_VA/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/exonsgenes.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bedtools('2019-7-26_cdna/hpmouse1/nc.flair',\n",
    "             'all_productive_isos.psl',\n",
    "            annot='software/flair/new.gencode.v29.annotation.gtf',\n",
    "            outfile='productive.annotated')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The number of transcripts accounted for in the output BAM generated by bedtools intersect is consistently smaller than the number of "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#debug code for dealing with isoforms intersecting with multiple genes\n",
    "\n",
    "outfile='2019-7-8_cdna/A431/trcl_VA/flair/isosproductive.annotated'\n",
    "reader=csv.reader(open(outfile),delimiter='\\t')\n",
    "transcript_dict={}\n",
    "\n",
    "for row in reader:\n",
    "    t=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "    name=row[17].split('gene_name \"')[1].split('\";')[0]\n",
    "    \n",
    "    if t not in transcript_dict:\n",
    "        transcript_dict[t]=[]\n",
    "        transcript_dict[t].append(name)\n",
    "    if t in transcript_dict:\n",
    "        if name not in transcript_dict[t]: transcript_dict[t].append(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attempt to debug transcripts intersecting multiple genes\n",
    "\n",
    "file=\"2019-6-25_dirrna/trcl_VA/flair/isoforms.annotated\"\n",
    "f=open(file)\n",
    "        \n",
    "for x in overlap_transcripts:\n",
    "    print x\n",
    "    counts={}\n",
    "    f.seek(0)     ##Resets reader to line 0\n",
    "    reader=csv.reader(f,delimiter='\\t')\n",
    "    for row in reader:\n",
    "        transcript=row[8].split('transcript_id \"')[1].split('\"')[0]\n",
    "        gene=row[17].split('gene_name \"')[1].split('\"')[0]\n",
    "        if transcript in x:\n",
    "            if gene in counts:\n",
    "                counts[gene]+=1\n",
    "            else:\n",
    "                counts[gene]=1\n",
    "    print counts\n",
    "        \n",
    "    bestscore=0\n",
    "    bestgene=''\n",
    "    for gene in counts:\n",
    "        if counts[gene]>bestscore:\n",
    "            bestgene=gene\n",
    "            bestscore=counts[gene]\n",
    "    print x+', best gene: '+bestgene+' '+str(bestscore)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Other notes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Other notes\n",
    "\n",
    "#Created fasta reference genome for only human chr.21\n",
    "!sed -n \"/^>NC_000021.9$/,/^>/p\" software/homo_GRCh38_trimmed_ref.fa >  software/homo_chr21.GRCh38_trimmed_ref.fa\n",
    "\n",
    "#Created human GENCODE annotation file of only exon and gene entries, to use for bedtools intersect\n",
    "!grep '\tgene\t' software/flair/new.gencode.v29.annotation.gtf > software/flair/genes.v29.annotation.gtf \n",
    "!grep '\texon\t' software/flair/new.gencode.v29.annotation.gtf > software/flair/exons.v29.annotation.gtf\n",
    "!cat software/flair/exons.v29.annotation.gtf software/flair/genes.v29.annotation.gtf > software/flair/exonsgenes.v29.annotation.gtf\n",
    "\n",
    "\n",
    "#Did not get around to using these 2 flair scripts but could be of good use for \n",
    "    ##identifying differential isoform expression\n",
    "    ##1)\n",
    "count_file='2019-7-26_cdna/a431.a1/trcl_VA/flair/counts_matrix.tsv'\n",
    "out_dir='2019-7-26_cdna/a431.a1/trcl_VA/flair/diffexp'\n",
    "!python software/flair/flair.py diffExp -q {count_file} -o {out_dir}\n",
    "    ##2)\n",
    "#python diff_iso_usage.py isoforms.psl colnum1 colnum2 diff_isos.txt\n",
    "isoforms='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/3.flair.collapse.isoforms.psl' \n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/diff_isos.txt'\n",
    "!python software/flair/bin/diff_iso_usage.py {isoforms} {outfile}\n",
    "\n",
    "#Bedtools annotate function, did not look into\n",
    "annot='software/flair/exonsgenes.v29.annotation.gtf'\n",
    "isoforms='2019-7-8_cdna/A431/trcl_VA/flair/3.flair.collapse.isoforms.gtf'\n",
    "output=\"2019-7-8_cdna/A431/trcl_VA/flair/bedtools.annotate\"\n",
    "!bedtools annotate -i {annot} -files {isoforms} > {output}\n",
    "\n",
    "#This flair script didn't do anything, just reappended on the chromosome position\n",
    "isoforms='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.collapse.isoforms.psl' \n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.named'\n",
    "annot='software/flair/new.gencode.v29.annotation.gtf'\n",
    "!python software/flair/bin/identify_gene_isoform.py {isoforms} {annot} {outfile}\n",
    "\n",
    "#Flair script identifies gene annotation for each entry of PSL. Only corrected a fraction of the isoforms\n",
    "    ##missing genes annotations\n",
    "\n",
    "psl='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/exonsgenes.collapse.isoforms.psl'\n",
    "ref='software/flair/new.gencode.v29.annotation.gtf'\n",
    "outfile='2019-7-8_cdna/SHSY5Y/trcl_VA/flair/isos_matched.psl'\n",
    "!python software/flair/bin/identify_annotated_gene.py {psl} {ref} {outfile}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
